src/mscorlib/src/System/Double.cs

```diff
@@ -44,56 +44,87 @@ public struct Double : IComparable, IFormattable, IConvertible
 
         internal static double NegativeZero = BitConverter.Int64BitsToDouble(unchecked((long)0x8000000000000000));
 
+        /// <summary>Determines whether the specified value is finite (zero, subnormal, or normal).</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
+        public unsafe static bool IsFinite(double d)
+        {
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) < 0x7FF0000000000000;
+        }
+
+        /// <summary>Determines whether the specified value is infinite.</summary>
+        [Pure]
+        [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
         public unsafe static bool IsInfinity(double d)
         {
-            return (*(long*)(&d) & 0x7FFFFFFFFFFFFFFF) == 0x7FF0000000000000;
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) == 0x7FF0000000000000;
         }
 
+        /// <summary>Determines whether the specified value is NaN.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
-        public static bool IsPositiveInfinity(double d)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
+        public unsafe static bool IsNaN(double d)
         {
-            //Jit will generate inlineable code with this
```

Comment:
The JIT seems to special case comparisons against PositiveInfinity and NegativeInfinity, will fix this to be `return (d == double.PositiveInfinity)`